### DISCUSSION TITLE
Best practice for unitialized venv

### DISCUSSION THREAD

**Discussion by manoelpqueiroz** [pyinvoke/invoke#1005] [Q&A]
Hi guys, I'm in quite a pickle attempting to move away from `Makefile` and use Invoke for setting up tasks. I have a repo where I want to make it easier for developers to set up the environment and start working with the best tools available, which would translate basically to:

1. Clone the repo;
2. Install Poetry for dependency management;
3. Use Poetry to create a virtual environment and install dependencies (which include Invoke as a dev dependency);
4. Start working! 😃

Since this is a project aimed at beginner devs, I would like to use Invoke to make it easier for them to execute steps 2 and 3. However, this poses a problem, because Invoke is being declared as a dev dependency (say in a `pyproject.toml` file) and at the same time I'm trying to create a task to download and install Poetry itself:

```py
from pathlib import Path
from invoke import Context, task

# Windows/Unix differentiation
BIN_DIR = "bin" if os.name != "nt" else "Scripts"
PTY = os.name != "nt"

# Virtualenv retrieval
ACTIVE_VENV = Path(os.environ.get("VIRTUAL_ENV", None))
VENV_HOME = Path(os.environ.get("WORKON_HOME", "~/virtualenvs"))
VENV_PATH = ACTIVE_VENV if ACTIVE_VENV.exists() else (VENV_HOME / PACKAGE_NAME)
VENV = VENV_PATH.expanduser()

VENV_BIN = Path(VENV) / Path(BIN_DIR)

PYTHON_PATH = VENV_BIN / "python"

@task
def poetry_download(c: Context) -> None:
    """Install Poetry as a standalone package via cURL.

    As per Poetry documentation, it should be either installed through a Python
    script which creates a temporary venv and installs Poetry in a
    platform-specific path or via pipx.

    More information can be obtained via https://python-poetry.org/docs/#installation
    and the script instructions in https://install.python-poetry.org/.

    """
    c.run(f"curl -sSL https://install.python-poetry.org | {PYTHON_PATH} -", pty=PTY)
```

If Invoke is not already installed as a system-wide package, this approach won't work. So far, I can only see two options:

- Resort to `Makefile` again to download Poetry and install dependencies, which seems counterintuitive since I'm aiming to move away from `Makefile` and it also won't work on Windows machines;
- Resign to only using Invoke for tasks after Poetry installation, which would require making it explicit in `README.md` or other documentation and might also be a source of more confusion to newcomers.

Do you guys see any other possible approach which allows for Invoke to handle everything? Or perhaps a best alternative if not possible?

**Comment by kuwv** [comment]
@manoelpqueiroz

You are mostly correct. There are three primary scopes to an installation: system, global, and local. The system scope is often maintained by native package managers. The global scope can be thought of as your home directory and the local being the directory path of your project.

Poetry itself needs to be installed and used either in the system or global scope. To automate poetry install with invoke you would need invoke (and the task library) preinstalled to be at that level or higher to install poetry. The only way to really accomplish this more efficiently would be to include invoke with Python or Poetry.

I have been aware of this for a while. I went with git hooks to run additional bootstrap the initialization tasks.

**Reply by manoelpqueiroz** [reply]
Hi @kuwv, I have also been thinking about this and it's a classic chicken and egg problem.

The git hook solution is nice! I ended up going for a more specific behaviour in my case with Cookiecutter: warn the user if invoke is unable to be imported after downloading the template and prompt them first to install via package manager if available.

**Comment by aucampia** [comment]
I would look into `uv run` with inline script metadata:

- https://docs.astral.sh/uv/guides/scripts/#running-a-script-with-dependencies
- https://packaging.python.org/en/latest/specifications/inline-script-metadata/
